### PR TITLE
ENH: add pyepics callback in background thread

### PR DESCRIPTION
**Pull request by ZLLentz**
Simple idea I had while ruminating over this slack thread: https://slac.slack.com/archives/C6B8GRCPQ/p1652395534927249

Since this can block up to 5 seconds, just do it in a thread maybe?
Does this have any terrible downstream effects? I'm not sure, but my uis are opening notably faster.

### CODE REVIEW

**Comment by klauer** [comment]
I think this makes a lot of sense. I think I would avoid a new thread for _every_ connection, though. 

I understand that my following suggestion will complicate the implementation, but I think it may be worth it in this case. Consider instead a thread pool of a certain (configurable?) size could potentially increase performance without taxing the CPU unnecessarily. For example, 4 threads that feed from a queue of PVs to monitor.

File: pydm/data_plugins/epics_plugins/pyepics_plugin_component.py [hunk 1]
```diff
@@ -47,6 +45,13 @@ def __init__(self, channel, pv, protocol=None, parent=None):
         self._upper_warning_limit = None
         self._lower_warning_limit = None
 
+        thread = CAThread(target=self.setup_callbacks, args=(channel,))
```
Comment (klauer) [inline pydm/data_plugins/epics_plugins/pyepics_plugin_component.py:48]:
```suggestion
        thread = CAThread(target=self.setup_callbacks, args=(channel,), daemon=True)
```
`daemon=True` to not let the thread running stop us from quitting pydm

**Comment by ZLLentz** [comment]
Yeah, thread pool makes sense- of course, that takes more than the 5 minutes of work that went into this implementation. Can come back to it on Monday.

Note that this isn't doing the monitor in a thread- it's just adding the callbacks, which is where pyepics is stalling out on us by doing a metadata get.

**Comment by jbellister-slac** [comment]
This looks like a good idea to me too. Gave it a spin on a few displays I use for testing and they all loaded up just fine.

**Comment by klauer** [comment]
Yes, definitely - didn't mean to diminish the work here. This is a good path forward! (And `monitor` was the wrong word choice above... `add_callback` to configure a monitor?)

**Comment by ZLLentz** [comment]
I ended up setting up the thread pool, it wasn't that hard

**Comment by jbellister-slac** [comment]
I am getting this while running some examples:

`Error in atexit._run_exitfuncs:
TypeError: shutdown() got an unexpected keyword argument 'cancel_futures'`

Looks like cancel_futures isn't available until python 3.9. Updating to 3.9 does clear it up, but need to be compatible with at least 3.7 in the mean time.

**Comment by ZLLentz** [comment]
Gotcha, I'll figure out what it needs to be in py3.7/3.8
One other thing: is there a better place to initialize the thread pool? I'm not sold that the module here is the best place to maintain the global state, but I also don't know the correct place to maintain the pyepics plugin's global state.

**Comment by jbellister-slac** [comment]
I have a bit of the same feeling, but looking around I don't see a place that makes more sense to put it. It's at least very clear what is going on by having it in this file, and I would be ok with leaving it here.

**Comment by ZLLentz** [comment]
I realize now that `PyEpicsPlugin` is a singleton that is created once- the correct place for this is on that class. Small adjustment.
Ken pointed out that the shutdown process of the threadpoolexecuter module does join all the active threads- this could potentially cause a slow shutdown if any long-running tasks are submitted to the threadpool. I suspect this may not matter much because the longest task we expect to see is ~5s and during startup, but it's worth thinking about...

**Review by klauer — APPROVED** [review]
This looks good to me. I think this is about as good as you can get with the limitations of the standard `ThreadPoolExecutor`.

Any improvements would have to come with a roll-your-own set of threads + a callback setup queue. If shutdown-related speed issues ever become a significant concern, this could be revisited.

**Review by jbellister-slac — APPROVED** [review]

**Pull request merged**